zerver/tests/test_events.py

```diff
@@ -4357,6 +4245,122 @@ def test_notification_setting_event_not_sent(self) -> None:
             )
         check_user_settings_update("events[0]", events[0])
 
+    def test_group_permission_update_events_on_user_deactivation(self) -> None:
+        othello = self.example_user("othello")
+        iago = self.example_user("iago")
+        self.user_profile = iago
+
+        realm = self.user_profile.realm
+        stream_name = "whatever"
+        stream = self.make_stream(stream_name, realm)
+
+        moderators_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MODERATORS, realm=realm, is_system_group=True
+        )
+        nobody_group = NamedUserGroup.objects.get(
+            name=SystemGroups.NOBODY, realm=realm, is_system_group=True
+        )
+
+        hamlet = self.example_user("hamlet")
+        setting_group = self.create_or_update_anonymous_group_for_setting(
+            [hamlet], [moderators_group]
+        )
+        do_change_realm_permission_group_setting(
+            realm,
+            "can_create_groups",
+            setting_group,
+            acting_user=None,
+        )
+        setting_group = self.create_or_update_anonymous_group_for_setting([hamlet], [])
+        do_change_realm_permission_group_setting(
+            realm,
+            "can_create_public_channel_group",
+            setting_group,
+            acting_user=None,
+        )
+
+        hamlet.refresh_from_db()
+        self.user_profile.refresh_from_db()
+
+        setting_group_member_dict = UserGroupMembersData(
+            direct_members=[hamlet.id, othello.id], direct_subgroups=[]
+        )
+        do_change_stream_group_based_setting(
+            stream,
+            "can_remove_subscribers_group",
+            setting_group_member_dict,
+            acting_user=iago,
+        )
+
+        setting_group_member_dict = UserGroupMembersData(
+            direct_members=[hamlet.id], direct_subgroups=[]
+        )
+        do_change_stream_group_based_setting(
+            stream,
+            "can_add_subscribers_group",
+            setting_group_member_dict,
+            acting_user=iago,
+        )
+
+        hamletcharacters_group = NamedUserGroup.objects.get(name="hamletcharacters", realm=realm)
+        setting_group = self.create_or_update_anonymous_group_for_setting(
+            [hamlet, othello], [moderators_group]
+        )
+        do_change_user_group_permission_setting(
+            hamletcharacters_group,
+            "can_mention_group",
+            setting_group,
+            acting_user=None,
+        )
+        setting_group = self.create_or_update_anonymous_group_for_setting([hamlet], [])
+        do_change_user_group_permission_setting(
+            hamletcharacters_group,
+            "can_manage_group",
+            setting_group,
+            acting_user=None,
```

Comment:
I think some comments would be helpful in all the test setup happening above. Hard to tell quickly why what setup is happening.